kolibri/utils/options.py

```diff
@@ -0,0 +1,103 @@
+import logging
+import os
+
+from configobj import ConfigObj
+from configobj import flatten_errors
+from configobj import get_extra_values
+from validate import Validator
+
+
+logger = logging.getLogger(__name__)
+
+ini_specification = [
+
+    "[Database]",
+    "DATABASE_ENGINE = option('sqlite', 'postgres', default='sqlite')",
+    "DATABASE_NAME = string(default='')",
+    "DATABASE_PASSWORD = string(default='')",
+    "DATABASE_USER = string(default='')",
+    "DATABASE_HOST = string(default='localhost')",
+
+    "[Paths]",  # these are defined relative to KOLIBRI_HOME (or absolute)
+    "CONTENT_ROOT = string(default=content)",
+
+    "[Deployment]",
+    "LISTEN_PORT = integer(default=8080)",
+
+]
+
+def get_option_name_list():
+
+    configspec = ConfigObj(ini_specification, _inspec=True)
+
+    option_names = []
+
+    for section in configspec.sections:
+        option_names += configspec[section].keys()
+
+    return option_names
+
+
+def read_options_file(ini_path):
+
+    ini_path = os.path.realpath(ini_path)
+
+    configspec = ConfigObj(ini_specification, _inspec=True)
+
+    conf = ConfigObj(ini_path, configspec=configspec, write_empty_values=True)
+
+    validation = conf.validate(Validator(), preserve_errors=True)
+
+    # loop over any errors with config values, and clean up
+    for section_list, key, error in flatten_errors(conf, validation):
+        if error is False:
+            error = 'missing value or section.'
+        if key is not None:
+            # remove the value that didn't validate, to ignore it (rather than bailing)
```

Comment:
The alternative (bailing) may be preferable.

Reply:
(Current behavior: output an error message, and then fall back to the default value, but keep running)

Reply:
Bailing might make issues easier to debug

Reply:
Yeah, it's a tradeoff, I guess -- if daemonized, could be hard to understand why the server stopped loading (e.g. after an upgrade -- although presumably we would want to always maintain back-compatibility with old ini files). I agree that bailing is probably "safest", though. Will wait for more opinions.